lib/chef/provider/user/mac.rb

```diff
@@ -0,0 +1,607 @@
+            end
+          end
+
+          requirements.assert(:all_actions) do |a|
+            a.assertion do
+              if new_resource.property_is_set?(:salt)
+                new_resource.property_is_set?(:iterations) && new_resource.password =~ /^[[:xdigit:]]{256}$/
+              else
+                true
+              end
+            end
+            a.failure_message(Chef::Exceptions::User, "Password must be a SALTED-SHA512-PBKDF2 shadow hash entropy when a shadow hash salt is given")
+          end
+        end
+
+        def load_current_resource
+          convert_group_name if new_resource.property_is_set?(:gid)
+
+          @current_resource = Chef::Resource::User::MacUser.new(new_resource.username)
+          current_resource.username(new_resource.username)
+
+          @user_plist = load_user
+          if user_plist
+            reload_with_user_plist
+          else
+            @user_exists = false
+            logger.trace("#{new_resource} user does not exist")
+          end
+
+          current_resource
+        end
+
+        def load_user
+          # Load the user information.
+          user_xml = run_dscl("read", "/Users/#{new_resource.username}")
+
+          # User doesn't exist yet so there's not info
+          return nil if user_xml.nil? || user_xml == ""
+
+          user_plist = Plist.parse_xml(user_xml)
+
+          # The password infomation is stored in the ShadowHashData key in the
+          # plist. However, parsing it is a bit tricky as the value is itself
+          # another encoded binary plist. We have to extract the encoded plist,
+          # decode it from hex to a binary plist and then convert the binary
+          # into XML plist. From there we can extract the hash data.
+          #
+          # NOTE: `dscl -read` and `plutil -convert` return different values for ShadowHashData.
+          #
+          # `dscl` returns the value encoded as an octal hex string and stored as a <string>
+          # `plutil` returns the value encoded as a base64 string stored as <data>
+          #
+          #  eg:
+          #
+          # <array>
+          #   <string>77687920 63616e27 74206170 706c6520 6275696c 6420636f 6e736973 74656e74 20746f6f 6c696e67</string>
+          # </array>
+          #
+          # vs
+          #
+          # <array>
+          #   <data>AADKAAAKAA4LAA0MAAAAAAAAAAA=</data>
+          # </array>
+          #
+          shadow_hash_hex = dscl_get(user_plist, :shadow_hash)[0]
+
+          return user_plist unless shadow_hash_hex && shadow_hash_hex != ""
+
+          shadow_binary_plist = [shadow_hash_hex.delete(" ")].pack("H*")
+          shadow_xml_plist = shell_out("plutil", "-convert", "xml1", "-o", "-", "-", input: shadow_binary_plist).stdout
+          dscl_set(user_plist, :shadow_hash, Plist.parse_xml(shadow_xml_plist))
+
+          user_plist
+        rescue Chef::Exceptions::PlistUtilCommandFailed, Chef::Exceptions::DsclCommandFailed
```

Comment:
it might be clearer to move these exceptions to wrap only the call which can throw them, because i think they're coming from different calls.  makes for more code, but it makes it clearer what can fail and where.